### ISSUE TITLE
edit during sourcing of vimrc: display issue with messages / missing undo

### ISSUE THREAD

**Issue by blueyed** [vim/vim#4984]
Given `t-edit-current-vimrc.vim`:
```vim
echom 1
echom 2
edit
echom 3
normal! cc
```

Running it (with `-u` and as argument) shows two issues:
```
% vim -Nu t-edit-current-vimrc.vim t-edit-current-vimrc.vim
1
2"t-edit-current-vimrc.vim" "t-edit-current-vimrc.vim" 5L, 40C
3
Press ENTER or type command to continue
```

1. messages are not handled correctly: the output from/for the edit should be on a new line probably, and also appears to have the new twice?

2. after continuing you see:
```
                                                                                    
echom 2
edit
echom 3
normal! cc
```

Here the first line it missing as expected (from the `cc`), but undoing (`u`) then results in an empty file (`5 fewer lines; before #1`).
Instead it should have an undo step for the `cc` there inbetween.

Not sure about this, and sorry for mixing issues - but I've noticed this while trying to debug undo issues with the SudoEdit plugin and thought it is worth reporting already.

Regarding 1. I've meant to reproduce an issue where echom-messages are hidden/overwritten completely due to `edit` being used in SudoEdit, which was/is very confusing while debugging.

Vim 8.1.1968

**Comment by brammool** [comment]
Daniel Hahler wrote:

> Given `t-edit-current-vimrc.vim`:
> ```vim
> echom 1
> echom 2
> edit
> echom 3
> normal! cc
> ```
> 
> Running it (with `-u` and as argument) shows two issues:
> ```
> % vim -Nu t-edit-current-vimrc.vim t-edit-current-vimrc.vim
> 1
> 2"t-edit-current-vimrc.vim" "t-edit-current-vimrc.vim" 5L, 40C
> 3
> Press ENTER or type command to continue
> ```
> 
> 1. messages are not handled correctly: the output from/for the edit
> should be on a new line probably, and also appears to have the new
> twice?

I don't see this problem, the "2" is on a line by itself.  Oh wait, with
"-N" it does happen.  It's because of the "o" flag in 'shortmess'.
Still, it should then overwrite the message, not append.  This can be
fixed with:

--- src/message.c       2019-09-21 20:46:14.736275700 +0200
+++ message.c   2019-09-26 20:51:37.566652624 +0200
@@ -1303,6 +1303,7 @@
 msg_start(void)
 {
     int                did_return = FALSE;
+    int                prev_msg_col = msg_col;
 
     if (!msg_silent)
        VIM_CLEAR(keep_msg);
@@ -1325,6 +1326,8 @@
            cmdmsg_rl ? Columns - 1 :
 #endif
            0;
+       if (msg_col == 0 && prev_msg_col != 0)
+           msg_putchar('\r');
     }
     else if (msg_didout)                   /* start message on next line */
     {

Please check if that helps or makes it worse.  Hmm, it does make several
tests fail.


> 2. after continuing you see:
> ```
>                                                                                     
> echom 2
> edit
> echom 3
> normal! cc
> ```
> 
> Here the first line it missing as expected (from the `cc`), but
> undoing (`u`) then results in an empty file (`5 fewer lines; before
> #1`).
> Instead it should have an undo step for the `cc` there inbetween.

The behavior is correct, this all becomes one undo block.  You can break
it by setting 'undolevels':
	set undolevels&

> Not sure about this, and sorry for mixing issues - but I've noticed
> this while trying to debug undo issues with the SudoEdit plugin and
> thought it is worth reporting already.
> 
> Regarding 1. I've meant to reproduce an issue where echom-messages are
> hidden/overwritten completely due to `edit` being used in SudoEdit,
> which was/is very confusing while debugging.

-- 
MORTICIAN:    Bring out your dead!
              [clang]
              Bring out your dead!
              [clang]
              Bring out your dead!
CUSTOMER:     Here's one -- nine pence.
DEAD PERSON:  I'm not dead!
                                  The Quest for the Holy Grail (Monty Python)

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by brammool** [comment]
Looks like this was already fixed by  8.1.2094.
Please close if you see it's fixed.

**Comment by blueyed** [comment]
Yes, it is fixed.
Thanks!
For reference: https://github.com/vim/vim/commit/473952e85286eb9c6098801f1819981ba61ad153#diff-3f7fe7da389a0156620b62301cfe8ad5R42-R63
